### PR TITLE
Move build system config into a separate module

### PR DESCRIPTION
**Pull request by snowleopard**
Another step towards splitting up the `build_system.ml` module. 

The first commit does the move, and the second commit simplifies `Rule_cache`. 

I'm planning to use this in the next PR that moves rule loading into separate module as well.

### CODE REVIEW

File: src/dune_engine/build_config.mli [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    -> t
+end
+
+type t =
+  { contexts : Build_context.t Context_name.Map.t Memo.Lazy.t
+  ; rule_generator : (module Rule_generator)
+  ; sandboxing_preference : Sandbox_mode.t list
+  ; mutable rule_done : int
+  ; mutable rule_total : int
+  ; mutable errors : Error.t list
```
Comment (rgrinberg) [inline src/dune_engine/build_config.mli:106]: It's not great to expose mutability in the mli. Also the name "config" seems strange to me. This record is basically the state of the build system, not some static part that was initialized.

Comment (snowleopard) [reply src/dune_engine/build_config.mli:106]: Yeah, I was thinking that we should perhaps split this into two parts: the immutable "config" and mutable state that will live in `build_system.ml`. Let me try to do that.

Comment (snowleopard) [reply src/dune_engine/build_config.mli]: OK, done.

**Review by rgrinberg — APPROVED** [review]

**Comment by snowleopard** [comment]
@rgrinberg Thanks for the review! I noticed that I missed `locks` and `build_mutex` and moved them to `State` too in the last commit, followed by a little clean up.

**Pull request merged**